libraries/standard/mqtt/src/mqtt.c

```diff
@@ -1374,81 +1508,109 @@ MQTTStatus_t MQTT_ProcessLoop( MQTTContext_t * pContext,
     uint32_t entryTimeMs = 0U, remainingTimeMs = timeoutMs, elapsedTimeMs = 0U;
     MQTTPacketInfo_t incomingPacket;
 
-    if( pContext != NULL )
+    if( ( pContext != NULL ) && ( pContext->callbacks.getTime != NULL ) )
     {
         getTimeStampMs = pContext->callbacks.getTime;
         entryTimeMs = getTimeStampMs();
         status = MQTTSuccess;
         pContext->controlPacketSent = false;
     }
-    else
+    else if( pContext == NULL )
     {
         LogError( ( "MQTT Context cannot be NULL." ) );
     }
+    else
+    {
+        LogError( ( "MQTT Context must set callbacks.getTime." ) );
+    }
 
     while( status == MQTTSuccess )
     {
-        status = MQTT_GetIncomingPacketTypeAndLength( pContext->transportInterface.recv,
-                                                      pContext->transportInterface.networkContext,
-                                                      &incomingPacket );
+        status = receiveSingleIteration( pContext, remainingTimeMs, true );
 
-        if( status == MQTTNoDataAvailable )
+        /* We don't need to break here since the status is already checked in
+         * the loop condition, and we do not want multiple breaks in a loop.
+         * Calculating remaining time before exiting is fine since it does not
```

Comment:
Even if Calculating remaining time before exiting is fine, is it needed or is it an unnecessary calculation. If latter is the case, would it be better to put it in the else block?

Reply:
It isn't needed, but it was this way so as not to increase complexity. My view was that the calculation is only unnecessary when the function returns an error, in which case spending a few extra cycles on a calculation would be the least concern. However, since some parts of this loop were moved to a separate function, the complexity's still ok with the additional nesting. I'm fine with moving the rest of the loop into an else block if you think that's better